### PR TITLE
broadcast: fix dictionary changed size during iteration error

### LINKED ISSUE #6222: Broadcast - GraphQLLocatedError: dictionary changed size during iteration
### Description

This is in CYLC_VERSION=8.3.2

```
2024-07-11T04:09:24Z INFO - Command "kill_tasks" actioned with 1 warnings. ID=baf57597-47c1-4ff5-a8a2-39875731e8e7
2024-07-11T04:09:25Z INFO - Command "broadcast" received.
    broadcast(cycle_points=['20230201T0000Z'], mode=put_broadcast, namespaces=['CLEAN_AND_RESET_002'], settings=[{'environment': {'FHRS': '000 012 024 036 042 048'}}])
2024-07-11T04:09:25Z INFO - Broadcast set:
    + [20230201T0000Z/CLEAN_AND_RESET_002] [environment]FHRS=000 012 024 036 042 048
2024-07-11T04:09:25Z INFO - [20230201T0000Z/archive_um_012_008/01:running] => succeeded
Traceback (most recent call last):
  File "miniconda3/envs/cylc-8.3.2/lib/python3.11/site-packages/promise/promise.py", line 844, in handle_future_result
    resolve(future.result())
            ^^^^^^^^^^^^^^^
  File "miniconda3/envs/cylc-8.3.2/lib/python3.11/site-packages/cylc/flow/network/graphql.py", line 433, in async_resolve
    return await next_(root, info, **args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "miniconda3/envs/cylc-8.3.2/lib/python3.11/site-packages/promise/iterate_promise.py", line 10, in iterate_promise
    yield from promise.future  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^
graphql.error.located_error.GraphQLLocatedError: dictionary changed size during iteration

2024-07-11T04:09:25Z ERROR - dictionary changed size during iteration
2024-07-11T04:09:25Z ERROR - graphql.error.located_error.GraphQLLocatedError: dictionary changed size during iteration
```

It appears to be a race condition with a dictionary not being threadsafe.

### Reproducible Example

I've not had time to explore this in great detail. I am roughly doing

* pause
* kill B (B is a family within A)
* broadcast to C (C is not related to A or B)  <---- this is where the error happens
* sleep 10
* hold A
* remove B (B is a family within A)
* sleep 10
* play
* trigger new flow from a specific task
* broadcast D (D is a family within A) to mimic skip mode (noting there are no xtriggers relating to D)
  e.g.
```
    cylc broadcast \
        -p "$CYLC_TASK_CYCLE_POINT" \
        -n "$namespace" \
        -s 'platform=localhost' \
        -s 'script=true' \
        -s 'pre-script=' \
        -s 'post-script=' \
        -s 'err-script=' \
        -s 'exit-script=' \
        "$CYLC_WORKFLOW_NAME"
 ```

* broadcast to C to remove the old broadcast
* release A

### Expected Behaviour

No failure should be seen.

### PR DESCRIPTION
**Pull request by oliver-sanders**
* Closes #6222

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [ ] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by oliver-sanders** [comment]
)Replicating the conditions that trigger this is rather problematic, hence the lack of a test)

**Review by hjoliver — APPROVED** [review]
Nice 👍

**Comment by hjoliver** [comment]
(I don't think this needs multiple reviewers, but since you've deliberately assigned several I'll leave it unmerged for now).

File: cylc/flow/data_store_mgr.py [hunk 1]
```diff
@@ -2251,7 +2251,9 @@ def delta_broadcast(self):
 
     def _generate_broadcast_node_deltas(self, node_data, node_type):
         cfg = self.schd.config.cfg
-        for node_id, node in node_data.items():
+        # NOTE: node_data may change during operation so make a copy
+        # see https://github.com/cylc/cylc-flow/pull/6397
+        for node_id, node in list(node_data.items()):
```
Comment (wxtim) [inline cylc/flow/data_store_mgr.py:2256]: Does this make a copy?

```python
>>> x = {1: 1, 2:2}
>>> x
{1: 1, 2: 2}
>>> y = x.items()
>>> y
dict_items([(1, 1), (2, 2)])
>>> x[1] = 2
>>> y
dict_items([(1, 2), (2, 2)])

```

Comment (wxtim) [inline cylc/flow/data_store_mgr.py:2256]: Does this work?

```python
>>> x = {"a": 1, "b": 2}
>>> for k, v in list(x.items()):
...   x[k] +=  1
... 
>>> x
{'a': 2, 'b': 3}
```

Comment (oliver-sanders) [reply cylc/flow/data_store_mgr.py:2256]: Try it in Python?

**Pull request merged**